cluster/v1alpha1/helpers.go

```diff
@@ -0,0 +1,222 @@
+		if status, exist := r.objs[cluster]; exist {
+			if status != UpgradeSucceed {
+				result = append(result, cluster)
+			}
+		}
+	}
+
+	return result
+}
+
+func (r *RolloutHandler) GetProgressiveClusters() []string {
+	// Progressive
+	if r.rolloutStrategy.Progressive == nil {
+		r.rolloutStrategy.Progressive = &RolloutProgressive{}
+	}
+
+	// upgrade madatory decision groups first
+	groupKeys := decisionGroupsToGroupKeys(r.rolloutStrategy.Progressive.MandatoryDecisionGroups.MandatoryDecisionGroups)
+	clusterGroupKeys, clusterGroups := r.pdTpdTracker.ExistingClusterGroups(groupKeys)
+
+	result := r.progressivePerGroup(clusterGroupKeys, clusterGroups, "None")
+	if len(result) > 0 {
+		return result
+	}
+
+	// upgrade the rest clusters
+	clusters := r.pdTpdTracker.Existing([]clusterv1beta1.GroupKey{})
+	length, _ := calculateLength(r.rolloutStrategy.Progressive.MaxConcurrency, len(clusters))
+	restclusters := r.pdTpdTracker.ExistingBesides(clusterGroupKeys).UnsortedList()
+
+	result = r.progressivePerCluster(restclusters, length, r.rolloutStrategy.Progressive.Timeout.Timeout)
+	return result
+}
+
+func (r *RolloutHandler) GetProgressivePerGroupClusters() []string {
+	// ProgressivePerGroup
+	if r.rolloutStrategy.ProgressivePerGroup == nil {
+		r.rolloutStrategy.ProgressivePerGroup = &RolloutProgressivePerGroup{}
+	}
+	// upgrade madatory decision groups first
+	groupKeys := decisionGroupsToGroupKeys(r.rolloutStrategy.Progressive.MandatoryDecisionGroups.MandatoryDecisionGroups)
+	clusterGroupKeys, clusterGroups := r.pdTpdTracker.ExistingClusterGroups(groupKeys)
+
+	result := r.progressivePerGroup(clusterGroupKeys, clusterGroups, "None")
+	if len(result) > 0 {
+		return result
+	}
+
+	// upgrade the rest decision groups
+	restClusterGroupKeys, restClusterGroups := r.pdTpdTracker.ExistingClusterGroupsBesides(clusterGroupKeys)
+
+	result = r.progressivePerGroup(restClusterGroupKeys, restClusterGroups, r.rolloutStrategy.ProgressivePerGroup.Timeout.Timeout)
+	return result
+}
+
+func (r *RolloutHandler) progressivePerCluster(clusters []string, length int, timeout string) []string {
+	result := []string{}
+	if length == 0 {
+		return result
+	}
+
+	sort.Strings(clusters)
+	for _, cluster := range clusters {
+		if status, exist := r.objs[cluster]; exist {
+			if canIAppendToResult(status, timeout) {
+				result = append(result, cluster)
+			}
+		} else {
+			result = append(result, cluster)
+		}
+
+		if (len(result)%length == 0) && len(result) > 0 {
+			return result
+		}
+	}
+
+	return result
+}
+
+func (r *RolloutHandler) progressivePerGroup(clusterGroupKeys []clusterv1beta1.GroupKey, clusterGroups map[clusterv1beta1.GroupKey]sets.Set[string], timeout string) []string {
+	result := []string{}
+
+	for _, key := range clusterGroupKeys {
+		if subclusters, ok := clusterGroups[key]; ok {
+			for _, cluster := range subclusters.UnsortedList() {
+				if status, exist := r.objs[cluster]; exist {
+					if canIAppendToResult(status, timeout) {
```

Comment:
canIAppendToResult is this actually isUpgrading?

Reply:
Not only is upgrading,  but also some failed but skipped clusters.

Reply:
May need a better function name.